src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -4719,11 +4719,15 @@ bool StorageReplicatedMergeTree::optimize(
         }
     }
 
-    if (query_context->getSettingsRef().replication_alter_partitions_sync != 0)
+    table_lock.reset();
+
+    Int64 wait_for_inactive_timeout = query_context->getSettingsRef().replication_wait_for_inactive_replica_timeout;
+    for (auto & merge_entry : merge_entries)
     {
-        /// NOTE Table lock must not be held while waiting. Some combination of R-W-R locks from different threads will yield to deadlock.
-        for (auto & merge_entry : merge_entries)
-            waitForAllReplicasToProcessLogEntry(merge_entry, false);
+        if (query_context->getSettingsRef().replication_alter_partitions_sync == 1)
```

Comment:
This code `if (replication_alter_partitions_sync == 1) else if  == 2` seems to be repeated 6-7 times. Maybe a common function would be better and goes in line with the changes in the PR to unify the behaviour.